### PR TITLE
Implement animations

### PR DESCRIPTION
**Pull request by romgrk**
Ok so I know this sounds crazy but it's not as bad as you think it is :laughing: I made two GIFs with different applications but both feel jankier than it feels in real. The blend (opacity) animation does feel nice, but I must admit the display-height animation feel a bit janky no matter what because it's limited to increasing by a line at each step :/

demo

clap-animate

(I only half-expect you to take this one seriously, I don't expect people to be enthusiastic with animations, feel free to decline ^^)

### CODE REVIEW

**Comment by liuchengxu** [comment]
Nice job! Looks super cool, but we need to ensure it won't introduce any performance regression before doing the merge.

File: autoload/clap/animate.vim [hunk 1]
```diff
@@ -0,0 +1,76 @@
+
+let g:clap_animation_frequency = get(g:, 'clap_animation_frequency', 50)
+
+function! clap#animate#start(duration, initial, final, type, Fn)
+  let duration = a:duration
+  let initial  = a:initial
+  let final    = a:final
+  let type     = a:type
```
Comment (liuchengxu) [inline autoload/clap/animate.vim:8]: This does not look very helpful, just using `a:*` is good.

File: autoload/clap/animate.vim [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+  call timer_stop(a:state.timer)
+endfunc
+
+function! clap#animate#lerp(ratio, initial, final, ...)
+  let type = a:0 > 0 ? a:000[0] : v:t_number
+
+  let range = a:final - a:initial
+  let delta = type == v:t_number ?
+    \ float2nr(a:ratio * range) :
+    \          a:ratio * range
```
Comment (liuchengxu) [inline autoload/clap/animate.vim:71]: This line is not too long, no need to wrap it?

```suggestion
  let delta = type == v:t_number ? float2nr(a:ratio * range) : a:ratio * range
```

File: autoload/clap/animate.vim (hunk 1 above)
Comment (romgrk) [reply autoload/clap/animate.vim:8]: Yes I was playing with the parameters format (between a dict or like that), I'll remove those.

**Comment by romgrk** [comment]
> Nice job! Looks super cool, but we need to ensure it won't introduce any performance regression before doing the merge.

Yes! This is really just a POC. The opacity animation works nice, but the height animation needs to be handled correctly. The height can be changed mid-animation by the `display.shrink*` methods, that's why I commented them for this. I'll need to model it that in a way that is clearer and ensure the animation adapts to any new requested height. I'll probably need to hide the `nvim_win_set_config(s:display_windi` calls behind a `s:display_set_config` that will check the animation status before applying the height. I'll complete the tags PR first though.

**Comment by HaleTom** [comment]
This looks very sweet!  Would you be up for looking at it again with fresh 2022 eyes?

**Comment by romgrk** [comment]
@HaleTom I don't have as much time for OSS these days :| Also I'm not interested enough in continuing this PR, I have other stuff to play with right now (integrating kitty graphics into neovim), but whoever wants to continue the work is welcome to take it.

**Comment by haskelcurry** [comment]
Hey just wanted to say it's awesome, thanks @romgrk

**Comment by romgrk** [comment]
Closing, stale.

**Pull request closed without merging**